core/src/apps/cardano/get_public_key.py

```diff
@@ -1,12 +1,77 @@
-from trezor.messages import CardanoVersion
-import apps.cardano.byron.get_public_key as byron
-import apps.cardano.shelley.get_public_key as shelley
+from ubinascii import hexlify
+
+from trezor import log, wire
+from trezor.messages.CardanoPublicKey import CardanoPublicKey
+from trezor.messages.HDNodeType import HDNodeType
+
+from apps.cardano import CURVE, seed
+from apps.common import HARDENED, layout, paths
+from apps.common.paths import is_hardened
+from apps.common.seed import remove_ed25519_prefix
 
 
 async def get_public_key(ctx, msg):
-    if msg.version is None or msg.version == CardanoVersion.BYRON:
-        return await byron.get_public_key(ctx, msg)
-    elif msg.version == CardanoVersion.SHELLEY:
-        return await shelley.get_public_key(ctx, msg)
+    # TODO is this what we want for public staking keys as well?
+    keychain = await seed.get_keychain(ctx, msg.address_n[:2])
+
+    await paths.validate_path(
+        ctx,
+        _validate_path_for_get_public_key,
+        keychain,
+        msg.address_n,
+        CURVE,
+        slip44_id=1815,
+    )
+
+    try:
+        key = _get_public_key(keychain, msg.address_n)
+    except ValueError as e:
+        if __debug__:
+            log.exception(__name__, e)
+        raise wire.ProcessError("Deriving public key failed")
+
+    if msg.show_display:
+        await layout.show_pubkey(ctx, key.node.public_key)
+    return key
+
+
+def _get_public_key(keychain, derivation_path: list):
+    node = keychain.derive(derivation_path)
+
+    public_key = hexlify(remove_ed25519_prefix(node.public_key())).decode()
+    chain_code = hexlify(node.chain_code()).decode()
+    xpub_key = public_key + chain_code
+
+    node_type = HDNodeType(
+        depth=node.depth(),
+        child_num=node.child_num(),
+        fingerprint=node.fingerprint(),
+        chain_code=node.chain_code(),
+        public_key=remove_ed25519_prefix(node.public_key()),
+    )
+
+    return CardanoPublicKey(node=node_type, xpub=xpub_key)
+
+
+def _validate_path_for_get_public_key(path: list, slip44_id: int) -> bool:
+    """
+    Checks if path has at least three hardened items and slip44 id matches.
+    The path is allowed to have more than three items, but all the following
+    items have to be non-hardened.
 
-    raise ValueError("Unsupported Cardano version '%s'" % msg.version)
+    Copied from apps.common.paths and modified to use 1852 instead of 44.
+    """
+    length = len(path)
+    if length < 3 or length > 5:
+        return False
+    if path[0] != 1852 | HARDENED:  # TODO put this constant in a single place
```

Comment:
This does not work for Byron public keys, does it?

Also, we might want to check if path[3] is appropriate (0/1 for Byron and 0/1/2 for Shelley). Hasn't this been done before in the Byron code?

Also, there is `validate_full_path`, and I don't like having two different functions for this, especially since I don't fully understand what is the intended difference. Let's discuss it on the call today --- are we supposed to derive *any* possible public key (with path belonging to the namespace), but do not allow address derivation along any path? (There were some limits on the number of accounts and addresses in Byron.) Or should the allowed paths be the same?

Reply:
This whole file isn't really completed yet. It's included in the PR because it's been moved. I will address the issues you mention later.

However, it's interesting, that the tests are not failing, even though the function actually returns false. Hm.

Not that I think about it, it makes sense to fix it in this PR. Byron functionality should not be broken and it wasn't broken before I moved the files. I'll fix it, thank you.